subprojects/docs/src/docs/userguide/feature_variants.adoc

```diff
@@ -0,0 +1,226 @@
+
+.Consuming specific features in a multi-project build
+====
+include::sample[dir="java-feature-variant/requiring-features/groovy/consumer",files="build.gradle[tags=consumer]"]
+include::sample[dir="java-feature-variant/requiring-features/kotlin/consumer",files="build.gradle.kts[tags=consumer]"]
+====
+
+Similarly, if the experimental Gradle metadata feature is enabled, it is possible to depend on an external dependency providing feature variants:
+
+.Consuming specific features from an external repository
+====
+include::sample[dir="java-feature-variant/requiring-features-external/groovy",files="build.gradle[tags=consumer]"]
+include::sample[dir="java-feature-variant/requiring-features-external/kotlin",files="build.gradle.kts[tags=consumer]"]
+====
+
+
+[[sec::incompatible_variants]]
+== Handling mutually exclusive variants
+
+The main advantage of using _capabilities_ as a way to handle features is that you can precisely handle compatibility of variants.
+The rule is simple: it's not allowed to have two variants of components that provide the same capability in a dependency graph.
+
+We can leverage that to ask Gradle to fail whenever the user mis-configures dependencies.
+Imagine, for example, that your library supports MySQL, Postgres and MongoDB, but that it's only allowed to choose _one_ of those at the same time.
+Not allowed should directly translate to "provide the same capability", so there must be a capability provided by all three features:
+
+.A producer of multiple features that are mutually exclusive
+====
+include::sample[dir="java-feature-variant/incompatible-variants/groovy/producer",files="build.gradle[tags=producer]"]
+include::sample[dir="java-feature-variant/incompatible-variants/kotlin/producer",files="build.gradle.kts[tags=producer]"]
+====
+
+Where, the producer declares 3 variants, one for each database runtime support:
+
+- `mysql-support` provides both the `db-support` and `mysql-support` capabilities
+- `postgres-support` provides both the `db-support` and `postgres-support` capabilities
+- `mongo-support` provides both the `db-support` and `mongo-support` capabilities
+
+Then if the consumer tries to get both the `postgres-support` and `mysql-support` like this (this also works transitively):
+
+.A consumer trying to use 2 incompatible variants at the same time
+====
+include::sample[dir="java-feature-variant/incompatible-variants/groovy/consumer",files="build.gradle[tags=consumer]"]
+include::sample[dir="java-feature-variant/incompatible-variants/kotlin/consumer",files="build.gradle.kts[tags=consumer]"]
+====
+
+trying to resolve would lead to an error:
+
+```
+Cannot choose between
+   org.gradle.demo:producer:1.0 variant mysqlSupportRuntimeElements and
+   org.gradle.demo:producer:1.0 variant postgresSupportRuntimeElements
+   because they provide the same capability: org.gradle.demo:producer-db-support:1.0
+```
+
+Because both variants provide the same capability.
```

Comment:
That last sentence feels a bit out of place at the bottom of the page. Since I read it as a repeat of info provided above, I would remove it.

Reply:
removed